src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -164,13 +165,33 @@ private Task StartAsyncCore()
                         _logger.HttpConnectionClosed();
                         return;
                     }
-
-                    _transport = _transportFactory.CreateTransport(GetAvailableServerTransports(negotiationResponse));
+                    _serverTransports = GetAvailableServerTransports(negotiationResponse);
+                    _transport = _transportFactory.CreateTransport(_serverTransports);
                     connectUrl = CreateConnectUrl(Url, negotiationResponse);
                 }
 
                 _logger.StartingTransport(_transport, connectUrl);
-                await StartTransport(connectUrl);
+                var validTransports = _requestedTransportType & _serverTransports;
```

Comment:
The bit-twiddling is clever, but I think I'd prefer something a little more clear-cut. My thought is that we have an ordered list of transport types internally, and we iterate through that. For each transport type we check it against the `_requestedTransportType` and `_serverTransports` masks to see if we should try it. If it passes that mask, we try it and if it fails we move on to the next one.

Something like:

```csharp
var allowedTransports = _requestedTransportType & _serverTransports;
foreach(var transport in KnownTransports)
{
    if ((transport & allowedTransports) != 0) {
        try {
            await StartTransport(connectUri, transport);
            // We succeeded, exit the loop
            break;
        } catch(Exception ex) {
            // Log and just allow the loop to continue.
        }
    }
}

// If we haven't started a transport yet, we fail here because we're out of transports to try.
```

Reply:
Much cleaner. This saves us from having to expose the transport type on ITransport as well.